### DISCUSSION TITLE
How can I import channels in a group, via commandline

### DISCUSSION THREAD

**Discussion by jerchap** [nextgenhealthcare/connect#5667] [Q&A]
Using the mirth CLI, I tried importing a group of channels that was previously exported

My commandline lokos like this:
`"c:\program files\mirth connect\mccommand.exe" -u admin -p admin -s commands.txt`

This is the contents of my commands.txt file:
`import c:\mirth\deploy\Mirth.RuleBuilder\Channels.xml force`

I just get an error "invalid channel file". If I import just a single channel without the group it works fine. I don't see any group related command in the help. Does this functionality exist?

**Comment by jonbartels** [comment]
When you get the error, what error is logged on the server in `mirth.log`?

**Reply by jerchap** [reply]
There is nothing in the log files.  

The command line response is:

```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.thoughtworks.xstream.core.util.Fields (file:/c:/program%20files/mirth%20connect/cli-lib/xstream-1.4.7.jar) to field java.util.TreeMap.comparator
WARNING: Please consider reporting this to the maintainers of com.thoughtworks.xstream.core.util.Fields
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Connected to Mirth Connect server @ https://127.0.0.1:8443 (3.9.1)
Executing statement: import c:\test\Channels.xml force
Error: invalid channel file.
Disconnected from server.
```

I think it's because the CLI is expecting 1 a file with a single channel (root element is <channel>) instead of a file that contains a group (root element <channelGroup>).  Too bad it's not supported, because we're looking to automate deployment to Mirth servers from Azure DevOps CI/CD pipelines, and we're so close.

**Reply by jonbartels** [reply]
You are correct. https://github.com/nextgenhealthcare/connect/blob/e0db5497aabb94c41071e4a098580e660a1d5d8e/command/src/com/mirth/connect/cli/CommandLineInterface.java#L291

That line is where the `import` command is parsed and there is no "import group" operation that I can see. 

Then here https://github.com/nextgenhealthcare/connect/blob/e0db5497aabb94c41071e4a098580e660a1d5d8e/command/src/com/mirth/connect/cli/CommandLineInterface.java#L1941 the XML for the import is parsed as Channel XML and not ChannelGroup XML.

The CLI is fairly well structured and you can see that it mostly just calls the MC `client` object to make API calls. If you're feeling feisty it shouldn't be too hard to alter the CLI and submit a PR. The `command` module should be compilable independently of the rest of the app.

**Reply by jerchap** [reply]
Thanks.  I might just do that.  Have not done any java in over a decade though :).  What's the IDE used, and is there any info out there on setting up the development environment?

**Reply by jonbartels** [reply]
https://github.com/nextgenhealthcare/connect/wiki/Development 

The best way to get started is fork the project, bang in some code even if it doesn't compile, and post a link here. 

Theres a fairly consistent flow in the CLI code:
1. A command flag like `importGroup`
2. The `switch` statement to call the import function
3. The actual import function will read the XML file and parse it into a ChannelGroup object
4. Use MC Client instance to POST the ChannelGroup object to the appropriate import/update method

Where it will be less obvious is:
1. Any pre-parsing
2. Does the MC API and MC Client take a ChannelGroup object to update or does it have to do some pre-processing and like push each channel then assign each channel to the group?

**Reply by bartha-adam** [reply]
Was looking for the same functionality of exporting/importing ChannelGroups...
